### DISCUSSION TITLE
How to deploy the bicep codes in Complete mode from azd?

### DISCUSSION THREAD

**Discussion by bioerrorlog** [Azure/azure-dev#2519] [Q&A] [answered]
Bicep's default deployment mode is Incremental mode, but I want to manage the resources in [Complete mode](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-modes#complete-mode).

In bicep, the `--mode` option can be used to specify the deoloyment mode, but how can I specify it in azd?

**Comment by rajyraman** [comment]
I am also interested in the answer to this question considering that we now have deployment stacks.

**Comment by vhvb1989** [comment]
I'm afraid `azd` supports `incremental` mode only (currently <= v1.1.0).

@savannahostrowski @ellismg @rajeshkamal5050.  We can easily add an argument to `provision` like `azd provision --mode=complete`, where the default mode is `incremental`.  
Another alternative is to use `azure.yaml` to set the `mode` within the `infra` field.  This approach would ensure the project is always using the same `mode`.

Let me know what you think.

**Reply by timkatje** [reply]
Glad someone already came up with this question and you guys are already thinking about it.  I know this would be extra work, but I think both of your proposed methods have great merit and that both should be tackled.

I often find that I need to run `complete` mode for one-offs to clean up something that got created by mistake, or that is left behind after an intended change.  For instance, I just renamed every resource to remove the resourceToken, which meant that I had two of every resource in my RG after running `azd up`.  In this situation, it would have been really convenient to have an argument to quickly get my RG back in a clean state.  Instead, the next quickest solution was to just `azd down` the project. :-(

However, there are certain projects where our deployment mode is always `complete` because they're sensitive environments that always need to be perfect.  For these projects, being able to set the mode in the `azure.yaml` file is ideal.  That way, when I pass the project off to my developers, it builds the environment exactly how I intended it.

One thing not mentioned, and `azd` may already do this by default, would be to permit the `--mode=complete` flag in any command that could invoke `provision`, i.e. `azd up --mode=complete`.  In that case, the argument name might need to be a little more precise, maybe something like `--armMode` or `--infra-mode`?  Or maybe just a boolean/swtich style argument like `--complete` or `--clean(up)`?

**Comment by WolfspiritM** [comment]
Coming from terraform I wonder why complete isn't even the default.

Lets say I have a prod environment deployed with azd.
With refactoring a service gets removed so it also gets removed from biceps.

Do I really have to manually delete that service?
What if company policy requires that all changes need to be done via IaC?

**Comment by IngerMathilde** [comment]
Any update regarding this feature? Currently this is blocking our company from fully using azd. Both having it as an argument `--mode=complete` and be able to set it in azure.yaml would have our personal preference, but either one only will also suffice.

**Comment by weikanglim** [comment, marked as answer]
With #4165 in latest release 1.10.1, we added deployment stacks support in `azd` as an alpha feature. If you enable this alpha feature, resources that you delete in Bicep files are automatically deleted in Azure when you run `azd provision`. Would this work for your scenario?

Note that this is still a slightly different behavior than Bicep `Complete` mode -- where resources in the *resource group* not referenced in the Bicep files are deleted automatically.

**Reply by rajeshkamal5050** [reply]
Bicep Complete mode support is tracked via https://github.com/Azure/azure-dev/issues/4241
